crates/miden-objects/src/block/nullifier_tree.rs

```diff
@@ -145,8 +147,10 @@ impl NullifierTree {
         nullifier: Nullifier,
         block_num: BlockNumber,
     ) -> Result<(), NullifierTreeError> {
-        let prev_nullifier_value =
-            self.smt.insert(nullifier.as_word(), Self::block_num_to_leaf_value(block_num));
+        let prev_nullifier_value = self
+            .smt
+            .insert(nullifier.as_word(), Self::block_num_to_leaf_value(block_num))
+            .map_err(|_| NullifierTreeError::NullifierPrefixAlreadyExists(nullifier))?;
```

Comment:
Apologies if I wasn't clear in my last comment. There can be multiple nullifiers in the same `SmtLeaf` so the error name is misleading. We just have to propagate the underlying error and so I would do:

```suggestion
        let prev_nullifier_value = self
            .smt
            .insert(nullifier.as_word(), Self::block_num_to_leaf_value(block_num))
            .map_err(|source| NullifierTreeError::MaxLeafEntriesExceeded { nullifier, source })?;
```